server/server/src/billing/billing_service.rs

```diff
@@ -379,113 +514,300 @@ pub async fn stripe_webhooks(
     match (&event.event_type, &event.data.object) {
         (stripe::EventType::InvoicePaymentFailed, stripe::EventObject::Invoice(invoice)) => {
             if let Some(stripe::InvoiceBillingReason::SubscriptionCycle) = invoice.billing_reason {
-                let customer_id = match invoice
-                    .customer
-                    .as_ref()
-                    .ok_or_else(|| {
-                        ClientError(StripeWebhookError::InvalidBody(
-                            "Cannot retrieve the customer_id.".to_string(),
-                        ))
-                    })?
-                    .deref()
-                {
-                    stripe::Expandable::Id(id) => id.to_string(),
-                    stripe::Expandable::Object(customer) => customer.id.to_string(),
-                };
-
-                let (public_key, user_info) =
-                    get_public_key_and_stripe_user_info(&event, &mut con, &customer_id).await?;
+                let public_key = get_public_key(&mut con, invoice).await?;
 
                 info!(
                     "User tier being reduced due to failed renewal payment via stripe. public_key: {}",
                     keys::stringify_public_key(&public_key)
                 );
 
-                con.set(data_cap(&public_key), FREE_TIER_USAGE_SIZE).await?;
-                con.json_set(stripe_user_info(&public_key), &user_info)
-                    .await?;
+                let tx_result = tx!(&mut con, pipe, &[keys::subscription_profile(&public_key)], {
+                    match lock_billing_workflow(
+                        &public_key,
+                        &mut con,
+                        server_state.config.stripe.millis_between_user_payment_flows,
+                    )
+                    .await
+                    {
+                        Ok(mut sub_profile) => {
+                            sub_profile.data_cap = FREE_TIER_USAGE_SIZE;
+                            sub_profile.billing_platform = None;
+
+                            pipe.json_set(keys::subscription_profile(&public_key), &sub_profile)
+                        }
+                        Err(ClientError(LockBillingWorkflowError::ExistingRequestPending)) => {
+                            Ok(&mut pipe)
+                        }
+                        Err(err) => Err(Abort(internal!(
+                            "Cannot get billing lock in stripe webhooks: {:#?}",
+                            err
+                        ))),
+                    }
+                });
```

Comment:
Do you tests cover this region of code? (I'm guessing no because webhooks)

and have you manually tested the stripe stuff? A lot of it has changed.

I have a suspicion that this region of code will infinite loop. Same with the one below it.

Reply:
long story short you're nesting transactions and you don't want to do this. Am surprised it works if it does.

Reply:
I will try to get rid of these nested transactions. Although it does work for android, I have not tested stripe, so I definitely need to check. I'll figure out a way to do this without the transaction business.

Reply:
Forgot to mention that my tests do not cover the google play webhook.